builder/googlecompute/step_start_tunnel.go

```diff
@@ -0,0 +1,323 @@
+	// Default: ".sh"
+	IAPExt string `mapstructure:"iap_ext" required:"false"`
+}
+
+type TunnelDriver interface {
+	StartTunnel(context.Context, string) error
+	StopTunnel()
+}
+
+func RunTunnelCommand(cmd *exec.Cmd) error {
+	// set stdout and stderr so we can read what's going on.
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Start()
+	if err != nil {
+		err := fmt.Errorf("Error calling gcloud sdk to launch IAP tunnel: %s",
+			err)
+		return err
+	}
+
+	// Give tunnel 30 seconds to either launch, or return an error.
+	// Unfortunately, the SDK doesn't provide any official acknowledgment that
+	// the tunnel is launched when it's not being run through a TTY so we
+	// are just trusting here that 30s is enough to know whether the tunnel
+	// launch was going to fail. Yep, feels icky to me too. But I spent an
+	// afternoon trying to figure out how to get the SDK to actually send
+	// the "Listening on port [n]" line I see when I run it manually, and I
+	// can't justify spending more time than that on aesthetics.
+	for i := 0; i < 30; i++ {
+		time.Sleep(1 * time.Second)
+
+		lineStderr, err := stderr.ReadString('\n')
+		if err != nil && err != io.EOF {
+			log.Printf("Err from scanning stderr is %s", err)
+			return fmt.Errorf("Error reading stderr from tunnel launch: %s", err)
+		}
+		if lineStderr != "" {
+			log.Printf("stderr: %s", lineStderr)
+		}
+
+		lineStdout, err := stdout.ReadString('\n')
+		if err != nil && err != io.EOF {
+			log.Printf("Err from scanning stdout is %s", err)
+			return fmt.Errorf("Error reading stdout from tunnel launch: %s", err)
+		}
+		if lineStdout != "" {
+			log.Printf("stdout: %s", lineStdout)
+		}
+
+		if strings.Contains(lineStderr, "ERROR") {
+			// 4033: Either you don't have permission to access the instance,
+			// the instance doesn't exist, or the instance is stopped.
+			// The two sub-errors we may see while the permissions settle are
+			// "not authorized" and "failed to connect to backend," but after
+			// about a minute of retries this goes away and we're able to
+			// connect.
+			// 4003: "failed to connect to backend". Network blip.
+			if strings.Contains(lineStderr, "4033") || strings.Contains(lineStderr, "4003") {
+				return RetryableTunnelError{lineStderr}
+			} else {
+				log.Printf("NOT RETRYABLE: %s", lineStderr)
+				return fmt.Errorf("Non-retryable tunnel error: %s", lineStderr)
+			}
+		}
+	}
+
+	log.Printf("No error detected after tunnel launch; continuing...")
+	return nil
+}
+
+type RetryableTunnelError struct {
+	s string
+}
+
+func (e RetryableTunnelError) Error() string {
+	return "Tunnel start: " + e.s
+}
+
+type StepStartTunnel struct {
+	IAPConf     *IAPConfig
+	CommConf    *communicator.Config
+	AccountFile string
+	ProjectId   string
+
+	tunnelDriver TunnelDriver
+}
+
+func (s *StepStartTunnel) ConfigureLocalHostPort(ctx context.Context) error {
+	if s.IAPConf.IAPLocalhostPort == 0 {
```

Comment:
@azr advice on the AWS SSM PR to pass LocalHostPort if set into the range config proved helpful. It is a nice way to ensure Packer retries if the port is currently in use.

Reply:
Yes, for example:

```go
		net.ListenRangeConfig{
			Min:     s.IAPConf.IAPLocalhostPort,
			Max:     s.IAPConf.IAPLocalhostPort,
```

Reply:
Nice. Just implemented. Once both are merged, I'll probably extract them into a common helper function since I think the code is nearly identical.